File: easy/0001-TwoSum/TwoSum.py
from typing import List

class Solution:
    # Time complexity: O(n)
    # Space complexity: O(n)
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        numsToIndex = {}
        for i in range(len(nums)):
            if target - nums[i] in numsToIndex:
                return [numsToIndex[target - nums[i]], i]
            numsToIndex[nums[i]] = i

    # Time complexity: O(n^2)
    # Space complexity: O(n)
    def twoSumBF(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if(nums[i] + nums[j] == target):
                    return [i, j]

solution = Solution()
print(solution.twoSum([2,7,11,15], 9))
print(solution.twoSum([3,2,4], 6))
print(solution.twoSum([3,3], 6))

File: easy/0001-TwoSum/TwoSum.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class TwoSum {
    // Time complexity: O(n)
    // Space complexity: O(n)
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> numsToIndex = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (numsToIndex.containsKey(target - nums[i])) {
                return new int[] {numsToIndex.get(target - nums[i]), i};
            }
            numsToIndex.put(nums[i], i);
        }
        throw new IllegalArgumentException();
    }

    // Time complexity: O(n^2)
    // Space complexity: O(n)
    public int[] twoSumBF(int[] nums, int target) {
        int[] indices = {};
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    indices = new int[] {i, j};
                    return indices;
                }
            }
        }
        return indices;
    }
    public static void main (String args[]) {
        TwoSum twoSum = new TwoSum();
        System.out.println(Arrays.toString(twoSum.twoSum(new int[]{2,7,11,15}, 9)));
        System.out.println(Arrays.toString(twoSum.twoSum(new int[]{3,2,4}, 6)));
        System.out.println(Arrays.toString(twoSum.twoSum(new int[]{3,3}, 6)));
    }
}